src/dshould/ShouldType.d

```diff
@@ -0,0 +1,218 @@
+module dshould.ShouldType;
+
+import std.algorithm : map;
+import std.format : format;
+import std.range : iota;
+import std.string : empty, join;
+import std.traits : TemplateArgsOf;
+public import std.traits : isInstanceOf;
+import std.typecons : Tuple;
+
+public auto should(T)(lazy T got) pure
+{
+    T get() pure @safe
+    {
+        return got;
+    }
+
+    return ShouldType!(typeof(&get))(&get);
+}
+
+public struct ShouldType(G, string[] words = [])
+{
+    import std.algorithm : canFind;
+
+    public G got;
+
+    private int* refs_ = null;
+
+    public auto addWord(string word)()
+    {
+        return ShouldType!(G, words ~ word)(this.got, this.refs);
+    }
+
+    private this(G got) { this.got = got; this.refs = 1; }
+
+    public this(G got, ref int refs) @trusted
+    in
+    {
+        assert(refs != CHAIN_TERMINATED, "don't copy Should that's been terminated");
+    }
+    body
+    {
+        this.got = got;
+        this.refs_ = &refs;
+        this.refs++;
+    }
+
+    this(this) @trusted
+    {
+        assert(this.refs != CHAIN_TERMINATED);
+
+        this.refs++;
+    }
+
+    ~this() @trusted
+    {
+        this.refs--;
+        assert(this.refs > 0, "unterminated should-chain!");
+    }
+
+    public enum gencheck(string TestString, string[] ArgNames) = gencheckImpl(TestString, ArgNames);
+
+    private static string gencheckImpl(string testString, string[] argNames)
+    {
+        auto args = argNames.join(", ");
+        auto argStrings = argNames.length.iota.map!(i => format!`"%s"`(argNames[i])).join(", ");
+
+        return format!q{{
+            import std.format : format;
+
+            check(mixin(format!q{%s}(%s)), format(": expected %s", %s), null, file, line);
+        }}(testString, argStrings, testString, args);
+    }
+
+    public void check(bool condition, lazy string left, lazy string right, string file, size_t line) pure @safe
+    {
+        terminateChain;
+
+        if (!condition)
+        {
+            throw new FluentException("test failed" ~ left, right, file, line);
+        }
+    }
+
+    public void allowOnlyWordsBefore(string[] allowedWords, string newWord)()
```

Comment:
For a reader of the function call it's not obvious that the second argument is only used for the error message.
The suffix "before" gives a hint in this function name, but it is missing in `requireWord`.
UFCS will not work. You could wrap the second argument like it's done for `approximately`:

    allowOnlyWords!(["be", "not"], before("empty"));
    requireWord!("be", before("empty"));

Reply:
Or else add the new word first: then there is no need to pass `words.back` as separate parameter?

Reply:
I don't understand your second comment.